presidio-analyzer/presidio_analyzer/predefined_recognizers/in_gstin_recognizer.py

```diff
@@ -0,0 +1,152 @@
+from typing import Optional, List, Tuple
+from presidio_analyzer import Pattern, PatternRecognizer
+from presidio_analyzer.analyzer_utils import PresidioAnalyzerUtils as Utils
+
+
+class InGstinRecognizer(PatternRecognizer):
+    """
+    Recognizes Indian Goods and Services Tax Identification Number ("GSTIN").
+
+    The GSTIN  is a fifteen character alpha-numeric code
+    with the last digit being a check digit calculated using a
+    modified modulus 36 LUHN calculation.
+    This recognizer identifies GSTIN using regex, context words and calculated value.
+    Reference: https://en.wikipedia.org/wiki/Goods_and_Services_Tax_(India),
+               http://idtc-icai.s3.amazonaws.com/download/knowledgeShare18-19/Structure-of-GSTIN.pdf
+
+    :param patterns: List of patterns to be used by this recognizer
+    :param context: List of context words to increase confidence in detection
+    :param supported_language: Language this recognizer supports
+    :param supported_entity: The entity this recognizer can detect
+    :param replacement_pairs: List of tuples with potential replacement values
+    for different strings to be used during pattern matching.
+    This can allow a greater variety in input, for example by removing dashes or spaces.
+    """
+
+    gstin_country_codes_iso3a = ""
+    utils = Utils()
+    countries = utils.get_country_codes(iso_code="ISO3166-1-Alpha-3")
```

Comment:
I'm trying to think of way to make this more efficient. Creating the `PresidioAnalyzerUtils` object for every recognizer that needs it is inefficient, and would require holding the countries data in memory multiple times.

On one hand, there's the need for this data for some recognizers, on the other hand most recognizers (and users) might not need it. Therefore, we can think of two options to handle this in my view.
1. Instantiate the `utils` in the `AnalyzerEngine` class and pass it to each recognizer. This would allow it to only be instantiated once.
2. Make some of these recognizers optional (which aren't loaded in `load_predefined_recognizers`) and therefore reduce the memory footprint. 

Long term, I think we should think of a mechanism to specify the countries you expect PII to come from. If someone doesn't expect any Australian license plates, it doesn't make sense to load that recognizer for them.

Happy to get your thoughts on this.

Reply:
hi @omri374 
I am confused with the best approach here frankly.  #2 is not a good idea as it will defeat the purpose of 'built-in' recognizers readily available as such from a consumer's perspective imho. 
We will need more of such metadata to improve detection beyond pattern matching as we add more industry specific recognizers  to give comprehensive solution offering.
So, how best to design this is bit complex ! 
#1 above seems to be a feasible approach to me but I will need fair bit of help here to implement since I don't want to end up breaking things due to my limited knowledge of the core. 
Shall we discuss offline ? Please suggest

Reply:
Hi @devopam. I agree that (1) is a better approach, also since in parallel we're thinking of introducing a country flag to be able to filter out recognizers that aren't needed.

Adding a new input to the `__init__` of `EntityRecognizer` would not break the API. I would suggest to do that, and instantiate the `PresidioAnalyzerUtils` in the recognizer registry, so that it could be passed to all recognizers during instantiation.

How about taking this constructor:
https://github.com/microsoft/presidio/blob/ea8d830fe445ee150e4ed76db5b4f36904400809/presidio-analyzer/presidio_analyzer/entity_recognizer.py#L35

and adding a new param to it:
```py
    def __init__(
        self,
        supported_entities: List[str],
        name: str = None,
        supported_language: str = "en",
        version: str = "0.0.1",
        context: Optional[List[str]] = None,
        analyzer_utils: Optional[PresidioAnalyzerUtils] = None
    ):
        self.analyzer_utils = analyzer_utils

```

We could inject the analyzer_utils during recognizers instantiation:
https://github.com/microsoft/presidio/blob/ea8d830fe445ee150e4ed76db5b4f36904400809/presidio-analyzer/presidio_analyzer/recognizer_registry.py#L139

WDYT?
Happy to discuss offline too.

Reply:
@devopam how about we have a quick chat on this? If you're interested, let's chat on LinkedIn

Reply:
Hi @omri374 ,
Please have a look at the current implementation and advise. Utils is instantiated only in the recognizers using it now viz. in_gstin, isin